src/connectors/tests/clickhouse/more_complex_test.rs

```diff
@@ -0,0 +1,404 @@
+// Copyright 2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The test in this module is very similar to what is done in `simple_test`.
+// The biggest difference is that we will try as much datatype as possible.
+
+// The table will contain the following columns:
+//
+// ---------------------------------
+// | Column name | Type            |
+// |-------------|-----------------|
+// | a           | UInt8           |
+// | b           | UInt16          |
+// | c           | UInt32          |
+// | d           | UInt64          |
+// | e           | Int8            |
+// | f           | Int16           |
+// | g           | Int32           |
+// | h           | Int64           |
+// | i           | String          |
+// | j           | DateTime        |
+// | k           | DateTime64(0)   |
+// | l           | DateTime64(3)   |
+// | m           | DateTime64(6)   |
+// | n           | DateTime64(9)   |
+// | o           | IPv4            |
+// | p           | IPv6            |
+// | q           | Uuid            |
+// | r           | Nullable(UInt8) |
+// | s           | Array(String)   |
+// ---------------------------------
+//
+// As the `DateTime64` encoding is not implemented in `clickhouse-rs`, the
+// columns `k`, `l`, `m` and `n` commented out everywhere in the file.
+
+use std::{
+    net::{Ipv4Addr, Ipv6Addr},
+    time::{Duration, Instant},
+};
+
+use chrono::{DateTime, NaiveDateTime, Utc};
+use chrono_tz::Tz;
+use clickhouse_rs::Pool;
+use testcontainers::{clients, core::Port, images::generic::GenericImage, RunnableImage};
+use tremor_common::ports::IN;
+use tremor_pipeline::{CbAction, Event, EventId};
+use tremor_script::literal;
+
+use super::utils;
+use crate::{
+    connectors::{
+        impls::clickhouse,
+        tests::{free_port, ConnectorHarness},
+    },
+    errors::{Error, Result},
+};
+
+const CONTAINER_NAME: &str = "clickhouse/clickhouse-server";
+const CONTAINER_VERSION: &str = "22.3.3.44";
+const SERVER_PORT: u16 = 9000;
```

Comment:
Would it make sense to move those one level up since they're shared between tests and currently redefined multiple times?

Reply:
It does. It is fixed by commit ccfb85c6.